firedrake/ml/pytorch/backend.py

```diff
@@ -0,0 +1,70 @@
+from firedrake.function import Function
+from firedrake.vector import Vector
+from firedrake.constant import Constant
+from firedrake.ml.backend_base import AbstractMLBackend
+
+import firedrake.utils as utils
+
+
+class PytorchBackend(AbstractMLBackend):
+
+    @utils.cached_property
+    def backend(self):
+        import torch
+        return torch
+
+    def custom_operator(self, *args, **kwargs):
+        from firedrake.ml.pytorch.pytorch_custom_operator import FiredrakeTorchOperator
+        return FiredrakeTorchOperator.apply(*args, **kwargs)
+
+    def to_ml_backend(self, x, gather=False, batched=True, **kwargs):
+        r"""Convert a Firedrake object `x` into a PyTorch tensor.
+
+            :arg x: Firedrake object (Function, Vector, Constant)
+            :kwarg gather: if True, gather data from all processes
+            :kwarg batched: if True, add a batch dimension to the tensor
+            :kwarg kwargs: additional arguments to be passed to torch.Tensor constructor
+                - device: device on which the tensor is allocated (default: "cpu")
+                - dtype: the desired data type of returned tensor (default: type of x.dat.data)
+                - requires_grad: if the tensor should be annotated (default: False)
+        """
+        if isinstance(x, (Function, Vector)):
+            if gather:
+                # Gather data from all processes
+                x_P = self.backend.tensor(x.vector().gather(), **kwargs)
+            else:
+                # Use local data
+                x_P = self.backend.tensor(x.vector().get_local(), **kwargs)
+            if batched:
+                # Default behaviour: add batch dimension after converting to PyTorch
+                return x_P[None, :]
+            return x_P
+        elif isinstance(x, Constant):
+            return self.backend.tensor(x.values(), **kwargs)
+        elif isinstance(x, (float, int)):
+            if isinstance(x, float):
+                # Set double-precision
+                kwargs['dtype'] = self.backend.double
+            return self.backend.tensor(x, **kwargs)
+        else:
+            raise ValueError("Cannot convert %s to a torch tensor" % str(type(x)))
+
+    def from_ml_backend(self, x, V=None):
```

Comment:
I guess if you are trying to implement some sort of ML interface then abstract methods might be useful. Have you considered using `@classmethod`?

Reply:
For mapping functions, currently, the instance only carries the backend but can also be used to access other information that might be added when utilising different machine learning frameworks or when GPU support is introduced.

For the PyTorch case, it is not needed for `from_ml_backend` since we use `torch.Tensor`'s sugar syntax to get the tensor in a suitable format for making the Firedrake objects. However, I am not sure that applies to other machine learning frameworks.

Reply:
I am wondering if one would ever want to instantiate more than one `PytorchBackend` instance. What would be the benefit?